### DISCUSSION TITLE
Is LDAP + TOTP possible?

### DISCUSSION THREAD

**Discussion by tonygermano** [kanidm/kanidm#2277] [Q&A]
I just found this project, and it looks really interesting. I've been looking for a directory with a self-service portal, the ability to authenticate LDAP clients, and perform MFA on the LDAP binds. I've been searching the docs and github, but did not find anything that addresses this specifically, and I'd like to know if it's possible.

I know both Okta and Jumpcloud support this by combining the TOTP code with the password when doing the bind (I remember this being a common strategy when adding MFA to RADIUS a long time ago, as well.)

**Comment by Firstyear** [comment]
Hi there,

You actually probably want something like https://github.com/kanidm/kanidm/issues/41 instead. 

The issue here is TOTP ... kinda sucks for user experience. And LDAP doesn't help that either, where we have to concat the password and the TOTP together to make it work. We'd rather not prioritise weird quirky hacks in LDAP, since as we push more with webauthn/passkeys, LDAP will not even be able to authenticate with this at all.

This is why we want to move to application passwords. You can have say an IMAP server setup to auth via ldap to an "IMAP specific" basedn. Then the user can configure their application password and when they are prompted they can just "copy-paste". This way you get MFA since to have access to that application password you must have used strong MFA to retrieve the password. The password can be stored long-term on a device (unlike LDAP + TOTP that would require re-entering the TOTP frequently as it changes). There can be many application passwords per user and application, so the users password on their phone is seperate to their laptop meaning you have specific device revocation and auditing. And finally, each application can then be seperated so that loss of a users IMAP password doesn't compromise adjacent services. 

Hope that helps,

**Reply by tonygermano** [reply]
Thank you for your prompt reply. I am doing some more investigation about the application passwords, but here is some more information about the problem I'm trying to solve.

I have 3 different but related legacy (windows) applications that can either use LDAP or each has their own internal user database. The usernames and passwords should be kept in sync between the 3 applications, as logging in to the first application passes the credentials and automatically logs in to the other two, so LDAP is my best option here.

In addition to the thick clients, the application servers have web portals exposed to the internet and contain sensitive information, so password + second factor is extremely important. Two of the applications have their own MFA implementations (which would be separate TOTP codes) and the 3rd does not support MFA at all. This is the reason that I wanted a single TOTP code integrated with LDAP for all 3 applications.

The credentials need to be entered on every login because the users frequently access the applications and services from shared workstations under the same user profile. Plus, the applications intentionally do not allow saving credentials.

**Comment by Firstyear** [comment]
> I have 3 different but related legacy (windows) applications that can either use LDAP or each has their own internal user database. The usernames and passwords should be kept in sync between the 3 applications, as logging in to the first application passes the credentials and automatically logs in to the other two, so LDAP is my best option here.

If you were to add TOTP here then the issue is that everything has to have the PW+TOTP passed around very quickly, because after 30 seconds it will no longer be valid and will cause new auth errors. So this sounds like it could be fragile.

Mostly what you are describing here sounds like what I could consider to be a "bad vendor implementation". If the application has a vendor or is opensource, you should raise issues about this since this kind of auth scheme is prone to all kinds of issues - the exact issues you are hitting here. Especially bad in this case is the inability to save the users credentials on a client, because that's the single feature here that would be blocking you from using this effectively. 

The other option for services like you are describing would be one where we have a TOTP-only ldap password, but the only way to retrieve the TOTP would be to login to the kanidm webui. That way you get MFA from the login to kani, then the user just needs to type in a 6 digit code to proceed.

**Reply by tonygermano** [reply]
I won't argue that it's bad vendor implementation 😆 These products are not open source, and I understand it would be fragile, but the credentials do get passed rather quickly from one app to the next (generally in under 2 seconds.) In the event that the password rotates between entering it and launching the child processes, the user would just need to manually authenticate in the child application.

As I mentioned, saving credentials on the client is not an option, because even if the applications were to allow it, the same client is shared by multiple users. They must authenticate and log out for each use.

Can you explain more about the TOTP-only ldap password? That sounds like it would be better than password-only. Why would that require logging into the kanidm webui to get the TOTP code rather than using an authenticator app?

**Reply by Firstyear** [reply]
There are a ton of considerations around this. We want to make attacks expensive, and we want to limit damage if a credential is exposed. 

In it's current form, the LDAP interface is just using the posix password - not great, because it's a single factor (something you know), it persists for a long time, and it is also the password used for sudo access - in future we want to start to reduce the LDAP interfaces ability to use this pw at all, reserving it for sudo only.

What you want here with LDAP PW + TOTP is the classic form of MFA - if an attacker gets the password then they *still* can't proceed because they lack the TOTP. It also means if the TOTP is disclosed then 30 seconds later it won't work anymore anyway. You need the token. 

Application passwords would isolate the password to each application - so one username + app specific password. Disclosure of one apps password only affects that one thing - you can't use the pw for sudo or logging into Kanidm or anything else. And to issue that password you need to have authenticated to Kanidm first with whatever account policy you may choose like MFA so the application password represents a "token" that you must have used mfa to get the app password in the first place. The drawback here is application passwords aren't meant for humans to memorise or enter, they are intended for storage in a devices password manager. 

The idea of the application TOTP is that you have a TOTP just for that one application. It's a 6 digit pin so it's easy for a human to type it out as needed, it's specific to the application so disclosure gets an attacker nothing. It also means you can't bruteforce it. 

The question then is do you just put the TOTP in the persons device/phone? In this scenario while the TOTP is a single factor, it's a very strong factor. Even in most systems people actually already consider the password compromised and the TOTP is the sole line of defence, so in that regard, why have the password at all? If the TOTP is doing all the heavy lifting security wise then we can rely on it as a very secure single factor!

Or you don't release the seed at all. The user needs to login to Kanidm with their webauthn keys or something similar, and then in the Kanidm webui that displays the current 6 digit application pin that the person needs. That way when the person uses that TOTP it's not just a TOTP, but a TOTP showing the user *must* have authenticated with MFA to access the TOTP in the first place. So the TOTP becomes a proof of MFA access, just like our application passwords were.

The pro of the former where the user puts the TOTP for the application into their device is convenience. They just pull our their phone or similar device and type in the PIN. The con is that the TOTP is single factor and loss of their device means instant access to the account.

The pro of the latter is that you get a strong guarantee the user logged in with MFA to Kanidm (via their phone even!) to get that TOTP. So even if they lose their phone, the attacker still can't access your application. The con is that it's more steps to *get* the TOTP to type in, where the user would need a yubikey or phone, they do a touchid/pin entry, they have to navigate the webui, then get the TOTP specific to that application. 

I think it's possible we could do this as a second stage of application passwords, but I'm not sure which approach we'd want. I suspect given we are already comfortable to store an application password on a device as a "single factor" then the TOTP on a single device is likely a similar security situation.

**Reply by yaleman** [reply]
TOTP and password per application is a bit of a usability nightmare, though, 😱 

Most password managers would make it even worse without specific tweaks/configuration (where possible, it's not for iCloud keychain), since a lot of applications are hosted on the same base domain and the password managers match on that, instead of the *full* hostname.

**Reply by tonygermano** [reply]
Well, the idea of doing password+TOTP over LDAP is actually fairly common (see links below.) It doesn't work well with password managers, but the fact that you are using TOTP means that the password itself does not need to be very long or complex. The TOTP (something you have) provides most of the security, while the password (something you know) prevents someone picking up your phone from being able to easily log in (two factors.)

In my case, I don't care that LDAP is using the posix password, because I'm not going to use that password for anything besides authenticating LDAP.

// commercial solutions
// These all do push notifications to their own mobile apps, but they also support combining password+TOTP as an alternative.
// Jumpcloud and Okta have their own user directories, but Duo uses a LDAP proxy service that sits between the application and another LDAP server.
https://jumpcloud.com/support/configure-mfa-for-ldap 
https://help.okta.com/oie/en-us/content/topics/directory/ldap-interface-mfa.htm
https://duo.com/docs/ldap

// open source
// glauth does not have a good end-user self-service portal
// I need to try privacyidea. It may work using the ldap proxy, but it looks much more complex to configure.
https://github.com/glauth/glauth
https://github.com/privacyidea/privacyidea-ldap-proxy (possibly)

**Reply by yaleman** [reply]
Oh, TOTP + password is well known... I was more responding to the above "per application" version that @Firstyear suggested.

**Reply by Firstyear** [reply]
I'd advise avoiding privacyidea, it has a lot of security issues. 

Anyway, I'm honestly not sure how to make a *nice* solution here @yaleman. Some things play great with application passwords (mail.app + fastmail for example). Somethings are awful. The issue here is the *application* that @tonygermano is using and the environment sounds like it has shared logins to a single desktop meaning you really complicate things. So having a way for a user to authenticate to a specific application in a way that forces keyboard interaction adds an extra challenge here.

So that is what led me down the path of per-application TOTP - because at least if the user is forced to type something, at least it doesn't have to be *much* that they have to type. And TOTP is generally relied upon in MFA because the assumption is "the password is already stolen". When you extrapolate from that you get "why have the password at all?". So my thinking here is that TOTP may not be "as nice" as a hardcore-per-app password, but it's a huge step ahead of the classic "LDAP password go everywhere". 

Edited with more thoughts.

**Reply by yaleman** [reply]
Yeah, but why individual TOTP per application AND an individual password per application?


Comparing ease of use vs security.

| option                 | security | usability |
| --- | --- | --- |
| application passwords    | ok | ok |
| application passwords + TOTP per-user better | good | less good but OK | 
| app passwords + per-application TOTP seeds  | super secure | support nightmare

OR do you mean just TOTP per-app instead of the app password + core TOTP?

**Reply by tonygermano** [reply]
I do want to reiterate that I have not one, but three tightly integrated, locally installed, applications with interactive logins that allow credential passing between apps so that manual login is not necessary on all three apps. App-specific passwords are the opposite of what I want. I'm looking at LDAP so that they have a single set of credentials and password resets only need to happen in one place instead of 3.

The TOTP code is meant to be a second factor, not the only factor. As I mentioned above, the TOTP code is meant to protect against a compromised password, but the password (even a simple one) is meant to protect against someone picking up the unlocked phone with the TOTP code. We also need to "check the MFA box" for cyber insurance requirements, and it's much easier to actually have multiple factors than to try to justify why only one is sufficient.

Given the set of applications, I think LDAP password+TOTP is the best option I have for security, user friendliness, and compliance.

**Reply by Firstyear** [reply]
I don't want to provide posix pw's over ldap in future, which only makes this harder. 

The entire concept of stuffing TOTP's on the end of a PW is also a usability nightmare and my experience with other products has only highlighted that. 

I think the only way I'd go ahead here is an application-specific pw+totp combo that the user can set, but that's a lot of extra work. (When I say application specific I mean that you can configure your trio of applications to "be the same application" to Kanidm and that way they could all validate this). Even if I committed to developing this, it's not a priority right now so expect it to be at least 6 months away. 

I think what you actually need to do here is go back to your vendor and insist that they improve their authentication options because passing around passwords like this is not a viable way to proceed in 2023.

Sorry :(

**Reply by tonygermano** [reply]
It's three separate vendors in the healthcare sector, so I don't expect them to change any time soon :smile: 

Thank you for taking the time to listen my situation and respond. I understand your position on this not being a priority for you as it is a legacy authentication model.

Last night I started looking at possibly using Zitadel to manage users and credentials, then setting up apacheds for the LDAP server with some custom code that calls the Zitadel API for search and bind operations.

**Comment by Firstyear** [comment]
> It's three separate vendors in the healthcare sector, so I don't expect them to change any time soon 😄

Why is it that it's always healthcare that has the absolute worst IT security practices 🤣 I know there are vendors in other places with hardcoded system-admin credentials on every deployment that has ever been installed. 

> Thank you for taking the time to listen my situation and respond. I understand your position on this not being a priority for you as it is a legacy authentication model.

Well this is why we designed the application specific password feature to handle this kind of legacy-ldap auth situation in a secure way, but the fact you can't "save the password" on the device is the huge limitation here. If it wasn't for that, it's actually exactly what you want here.

**Reply by tonygermano** [reply]
I don't think any of these workstations are assigned 1:1 to users. The workstations are more tied to the shift, so many users end up using the same device. They are also spread out across many physical locations where the local networks are controlled by different entities, so the machines themselves are not joined to a windows domain and a generic user is used for logging into the operating system since local account management would be a nightmare. The interactive logins in the applications are what guard access to the data.

This is actually not that uncommon of a scenario in this field. We basically have to work from the assumption that the device is compromised since so many people have access to it, so device-based auth doesn't make sense.

**Reply by Firstyear** [reply]
I wonder if you could use yubikeys in the mode where they emit a static pw instead here? 🤔 You wouldn't even need kani then, just the key and it would be way better than human memorised pw's

**Comment by Firstyear** [comment]
> This is actually not that uncommon of a scenario in this field. We basically have to work from the assumption that the device is compromised since so many people have access to it, so device-based auth doesn't make sense.

Also worth pointing out, once you assume a device is compromised, then it doesn't matter what you do because they can just keylog anyones interactions. Device compromise == hard game over.

**Reply by tonygermano** [reply]
I made that sounds worse than it is. The generic user is not a local admin, and the workstations are in a more-or-less physically secure area. It's just that many legitimate users have access to and share the same user profile.

TOTP would be what would thwart the keylogger, anyway, unless the attacker was capturing and reusing credentials in real-time.

**Reply by Firstyear** [reply]
Most TOTP attacks are realtime attacks these days. It's a thing. ( https://phishdeck.com/blog/what-is-realtime-phishing ) That's why there is such a huge push to webauthn.

**Reply by tonygermano** [reply]
I can see phishing proxies being an issue for popular public web sites, but I don't think that would be a problem for me where the users are normally launching desktop shortcuts to start the applications.

I think there needs to be some agreement that best practice solutions that depend on the consumption device being secured by a single user wlll not be best practice when the consumption device is shared by and does not differentiate between multiple users. Those rely on the application rather than the device to interactively authenticate the user, and that situation will likely not be going away.

Interactively authenticating on a public device should involve multiple factors, which is likely going to be harder to improve upon than password + TOTP while remaining usable, as neither of those require special hardware to be present, since a user will almost always have their cell phone with them. Back in the days of RSA keyfobs it was easier to forget them.

Basically, I don't think it's a one-size-fits-all thing, and different situations call for different solutions.

**Reply by tonygermano** [reply]
And I do consider TOTP more secure than push notifications.

**Reply by Firstyear** [reply]
I guess the issue here is my goal was always to do "what was best" in kani, because ive seen too many projects half-ass things and then can never change later as suddenly "oh everything depends on passwords". I dont want to fall into that. I dont want to lock us into bad practice.

So as mentioned. You have vendors, who are half-assing their security and then expect you to consume auth providers that do things badly. I dont think its up to kani to implement features that keeps enabling these vendors to get away with malpractice. Especially in something as important as healthcare.

At the same time security compliance is widely regarded as a running joke, for obvious reasons because its practioners are preaching outdated rote advice that true experts have moved away from 20 years ago. Im surprised you arent being force to change passwords every 90 days too tbh. 

Today its easier than ever to have secure cryptographic auth between webauthn and most security keys implementing PIV.  TOTP by comparison is annoying for users who have to pull out their phone, unlock it, find the app, find their totp, type it out and then wait for the server to auth. TOTP, frankly, sucks. I have heard many stories of users who refuse to enable mfa because they hate TOTP so much.

So i think in this case we have to accept that kani isnt going to work for your applications, because the applications rely on insecure practices and your compliance requirements are asking for something that we dont want to support.

if your vendors update their systems to properly support things like PIV or OAUTH/Oidc, then inthink kani is a good fit. Otherwise, you may need to accept going with something else sorry. 

Good luck with the auth setup and the vendors :)

**Reply by tonygermano** [reply]
I was enjoying the back and forth security discussion and did not expect to persuade you to do anything differently than what you had already told me :)

Though, I did just find that in addition to Okta, Duo, and JumpCloud, OneLogin is another commercial solution that offers LDAP+TOTP.

**Reply by Firstyear** [reply]
Hey I'm always happy to chat about security - join us on our channel https://gitter.im/kanidm/community 

I think they offer it because they have to, not because it's good. I've worked with/on ldap for a long time and I don't want to be stuck or limited in it's ideas :)
